Migrations: Convert enum values to underlying type

When creating a migration to add an `enum` column, the namespace of the enum is not included in the resulting migration.

EG: 

```
  migrationBuilder.AddColumn<int>(
                name: "Status",
                table: "CommandRecord",
                nullable: false,
                defaultValue: CommandStatus.Pending);
```

This results in `DNX,Version=v4.5.1 error CS0103: The name 'CommandStatus' does not exist in the current context` when running `dnx ef database update`

--- a/src/Microsoft.EntityFrameworkCore.Relational/Metadata/RelationalPropertyAnnotations.cs
+++ b/src/Microsoft.EntityFrameworkCore.Relational/Metadata/RelationalPropertyAnnotations.cs
@@ -2,6 +2,7 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Reflection;
 using JetBrains.Annotations;
 using Microsoft.EntityFrameworkCore.Metadata.Internal;
 using Microsoft.EntityFrameworkCore.Utilities;
@@ -29,7 +30,8 @@ namespace Microsoft.EntityFrameworkCore.Metadata
         public virtual string ColumnName
         {
             get { return (string)Annotations.GetAnnotation(RelationalAnnotationNames.ColumnName) ?? Property.Name; }
-            [param: CanBeNull] set { SetColumnName(value); }
+            [param: CanBeNull]
+            set { SetColumnName(value); }
         }
 
         protected virtual bool SetColumnName([CanBeNull] string value)
@@ -38,7 +40,8 @@ namespace Microsoft.EntityFrameworkCore.Metadata
         public virtual string ColumnType
         {
             get { return (string)Annotations.GetAnnotation(RelationalAnnotationNames.ColumnType); }
-            [param: CanBeNull] set { SetColumnType(value); }
+            [param: CanBeNull]
+            set { SetColumnType(value); }
         }
 
         protected virtual bool SetColumnType([CanBeNull] string value)
@@ -47,7 +50,8 @@ namespace Microsoft.EntityFrameworkCore.Metadata
         public virtual string GeneratedValueSql
         {
             get { return (string)Annotations.GetAnnotation(RelationalAnnotationNames.GeneratedValueSql); }
-            [param: CanBeNull] set { SetGeneratedValueSql(value); }
+            [param: CanBeNull]
+            set { SetGeneratedValueSql(value); }
         }
 
         protected virtual bool SetGeneratedValueSql([CanBeNull] string value)
@@ -59,15 +63,24 @@ namespace Microsoft.EntityFrameworkCore.Metadata
             {
                 return Annotations.GetAnnotation(RelationalAnnotationNames.DefaultValue);
             }
-            [param: CanBeNull] set { SetDefaultValue(value); }
+            [param: CanBeNull]
+            set { SetDefaultValue(value); }
         }
 
         protected virtual bool SetDefaultValue([CanBeNull] object value)
         {
-            if ((value != null)
-                && (Property.ClrType.UnwrapNullableType() != value.GetType()))
+            if (value != null)
             {
-                throw new InvalidOperationException(RelationalStrings.IncorrectDefaultValueType(value, value.GetType(), Property.Name, Property.ClrType, Property.DeclaringEntityType.DisplayName()));
+                var valueType = value.GetType();
+                if (Property.ClrType.UnwrapNullableType() != valueType)
+                {
+                    throw new InvalidOperationException(RelationalStrings.IncorrectDefaultValueType(value, valueType, Property.Name, Property.ClrType, Property.DeclaringEntityType.DisplayName()));
+                }
+
+                if (valueType.GetTypeInfo().IsEnum)
+                {
+                    value = Convert.ChangeType(value, valueType.UnwrapEnumType());
+                }
             }
 
             return Annotations.SetAnnotation(RelationalAnnotationNames.DefaultValue, value);

--- a/test/Microsoft.EntityFrameworkCore.Commands.FunctionalTests/Migrations/ModelSnapshotTest.cs
+++ b/test/Microsoft.EntityFrameworkCore.Commands.FunctionalTests/Migrations/ModelSnapshotTest.cs
@@ -75,6 +75,23 @@ namespace Microsoft.EntityFrameworkCore.Commands.Migrations
         {
         }
 
+        private enum Days : long
+        {
+            Sun,
+            Mon,
+            Tue,
+            Wed,
+            Thu,
+            Fri,
+            Sat
+        }
+
+        private class EntityWithEnumType
+        {
+            public int Id { get; set; }
+            public Days Day { get; set; }
+        }
+
         #region Model
 
         [Fact]
@@ -693,6 +710,29 @@ builder.Entity(""Microsoft.EntityFrameworkCore.Commands.Migrations.ModelSnapshot
     });
 ",
                 o => { Assert.Equal(true, o.GetEntityTypes().First().FindProperty("AlternateId").IsConcurrencyToken); });
+        }
+
+        [Fact]
+        public void Property_default_value_of_enum_type_is_stored_in_snapshot_without_actual_enum()
+        {
+            Test(
+                builder => { builder.Entity<EntityWithEnumType>().Property(e => e.Day).HasDefaultValue(Days.Wed); },
+                @"
+builder.Entity(""Microsoft.EntityFrameworkCore.Commands.Migrations.ModelSnapshotTest+EntityWithEnumType"", b =>
+    {
+        b.ToTable(""EntityWithEnumType"");
+
+        b.Property<int>(""Id"")
+            .ValueGeneratedOnAdd();
+
+        b.Property<long>(""Day"")
+            .ValueGeneratedOnAdd()
+            .HasAnnotation(""Relational:DefaultValue"", 3L);
+
+        b.HasKey(""Id"");
+    });
+",
+                o => { Assert.Equal(3L, o.GetEntityTypes().First().FindProperty("Day")["Relational:DefaultValue"]); });
         }
 
         #endregion

--- a/test/Microsoft.EntityFrameworkCore.Relational.Tests/Metadata/RelationalBuilderExtensionsTest.cs
+++ b/test/Microsoft.EntityFrameworkCore.Relational.Tests/Metadata/RelationalBuilderExtensionsTest.cs
@@ -146,6 +146,23 @@ namespace Microsoft.EntityFrameworkCore.Metadata.Tests
         }
 
         [Fact]
+        public void Can_set_column_default_value_of_enum_type()
+        {
+            var modelBuilder = CreateConventionModelBuilder();
+
+            modelBuilder
+                .Entity<Customer>()
+                .Property(e => e.EnumValue)
+                .HasDefaultValue(MyEnum.Tue);
+
+            var property = modelBuilder.Model.FindEntityType(typeof(Customer)).FindProperty("EnumValue");
+
+            Assert.Equal(typeof(ulong), property.Relational().DefaultValue.GetType());
+            Assert.Equal((ulong)2, property.Relational().DefaultValue);
+            Assert.Equal(ValueGenerated.OnAdd, property.ValueGenerated);
+        }
+
+        [Fact]
         public void Can_set_key_name()
         {
             var modelBuilder = CreateConventionModelBuilder();
@@ -1000,10 +1017,18 @@ namespace Microsoft.EntityFrameworkCore.Metadata.Tests
             Assert.Same(typeof(int), sequence.ClrType);
         }
 
+        private enum MyEnum : ulong
+        {
+            Sun,
+            Mon,
+            Tue
+        }
+
         private class Customer
         {
             public int Id { get; set; }
             public string Name { get; set; }
+            public MyEnum EnumValue { get; set; }
 
             public IEnumerable<Order> Orders { get; set; }
         }

--- a/test/Microsoft.EntityFrameworkCore.Relational.Tests/Metadata/RelationalMetadataExtensionsTest.cs
+++ b/test/Microsoft.EntityFrameworkCore.Relational.Tests/Metadata/RelationalMetadataExtensionsTest.cs
@@ -162,6 +162,28 @@ namespace Microsoft.EntityFrameworkCore.Metadata.Tests
         }
 
         [Fact]
+        public void Can_get_and_set_column_default_value_of_enum_type()
+        {
+            var modelBuilder = new ModelBuilder(new ConventionSet());
+
+            var property = modelBuilder
+                .Entity<Customer>()
+                .Property(e => e.EnumValue)
+                .Metadata;
+
+            Assert.Null(property.Relational().DefaultValue);
+
+            property.Relational().DefaultValue = MyEnum.Mon;
+
+            Assert.Equal(typeof(byte), property.Relational().DefaultValue.GetType());
+            Assert.Equal((byte)1, property.Relational().DefaultValue);
+
+            property.Relational().DefaultValue = null;
+
+            Assert.Null(property.Relational().DefaultValue);
+        }
+
+        [Fact]
         public void Throws_when_setting_column_default_value_of_wrong_type()
         {
             var modelBuilder = new ModelBuilder(new ConventionSet());
@@ -590,11 +612,19 @@ namespace Microsoft.EntityFrameworkCore.Metadata.Tests
             Assert.Contains(sequences, s => s.Name == "Golomb");
         }
 
+        private enum MyEnum : byte
+        {
+            Son,
+            Mon,
+            Tue
+        }
+
         private class Customer
         {
             public int Id { get; set; }
             public string Name { get; set; }
             public Guid AlternateId { get; set; }
+            public MyEnum? EnumValue { get; set; }
         }
 
         private class SpecialCustomer : Customer

